runtime/src/Messaging.tsx

```diff
@@ -1,71 +1,60 @@
-const HEARTBEAT_INTERVAL = 60;
-
-let pubnub: PubNub;
+let transport: RuntimeTransport | null = null;
 
 // Device metadata that is sent with every message from us
-const device = {
+const device: Device = {
   id: '', // async, populated in init
   name: Constants.deviceName,
   platform: Platform.OS,
 };
 
-let currentChannel: string | null = null;
-
-export const init = (deviceId: string) => {
-  pubnub = new PubNub({
-    publishKey: 'pub-c-2a7fd67b-333d-40db-ad2d-3255f8835f70',
-    subscribeKey: 'sub-c-0b655000-d784-11e6-b950-02ee2ddab7fe',
-    uuid: JSON.stringify({
-      id: deviceId,
-      name: Constants.deviceName,
-      platform: Platform.OS,
-    }),
-    ssl: true,
-    presenceTimeout: PRESENCE_TIMEOUT,
-    heartbeatInterval: HEARTBEAT_INTERVAL,
-  });
+export const init = (deviceId: string, testTransport: string | null | undefined) => {
   device.id = deviceId;
+  let transportClass;
+  if (Platform.OS === 'web') {
+    transportClass = require('./transports/RuntimeTransportImplWebPlayer').default;
+  } else if (testTransport === 'snackpub') {
+    transportClass = require('./transports/RuntimeCompositedTransport').default;
+  } else {
+    transportClass = require('./transports/RuntimeTransportImplPubNub').default;
+  }
+  transport = new transportClass(device);
 };
 
-// End existing PubNub subscription, if any
 export const unsubscribe = () => {
-  if (currentChannel) {
-    Logger.comm('Unsubscribing from channel', currentChannel);
-    pubnub.unsubscribe({ channels: [currentChannel] });
+  if (!transport) {
+    throw new Error('Transport does not initialize. Please call `Messaging.init()` first.');
   }
+  transport.unsubscribe();
 };
 
-// Initiate PubNub subscription to given `channel`, throw if couldn't connect / timed out. Ends
-// existing subscription, if any.
 export const subscribe = ({ channel }: { channel: string }) => {
-  unsubscribe();
-
-  currentChannel = channel;
-  Logger.comm('Subscribing to channel', channel);
-  pubnub.subscribe({ channels: [channel] });
+  if (!transport) {
+    throw new Error('Transport does not initialize. Please call `Messaging.init()` first.');
+  }
+  transport.subscribe(channel);
 };
 
-// Add a message listener
-export const listen = (listener: ListenerParameters['message']) => {
-  pubnub.addListener({ message: listener });
+export const listen = (listener: (payload: RuntimeMessagePayload) => void) => {
+  if (!transport) {
+    throw new Error('Transport does not initialize. Please call `Messaging.init()` first.');
+  }
+  transport.listen(listener);
 };
 
-// Publish a message to the currently subscribed channel, if any
 export const publish = (message: object) => {
-  if (currentChannel) {
-    Logger.comm('Sending message', message);
-    pubnub.publish({
-      channel: currentChannel,
-      message: { ...message, device },
-    });
+  if (!transport) {
+    throw new Error('Transport does not initialize. Please call `Messaging.init()` first.');
```

Comment:
```suggestion
    throw new Error('Transport not initialized, call `Messaging.init()` first.');
```

Reply:
Also, do we need to throw or can we log this somewhere? Not sure how useful it is if our users are running into this exception?

Reply:
> Also, do we need to throw or can we log this somewhere? Not sure how useful it is if our users are running into this exception?

this is developer error and should not happen. somehow to make typescript happy like `nullthrow(transport)` or `assert(transport)`. since it's running in app runtime, i don't want to introduce more dependencies for nullthow or assert.

Reply:
That makes sense 👍